### ISSUE TITLE
RedisArray and sscan

### ISSUE THREAD

**Issue by akalongman** [phpredis/phpredis#2055]
Looks like this bug is related to RedisArray class https://github.com/laravel/framework/issues/40149

cc @yatsukhnenko

**Comment by yatsukhnenko** [comment]
@akalongman could you provide a script or dockerfile to reproduce the issue?

**Comment by akalongman** [comment]
@yatsukhnenko 

```php
<?php
ini_set('error_reporting', E_ALL);
ini_set('display_errors', 1);

$servers = [
    '127.0.0.1:6379',
    '127.0.0.1:6380',
];

$client = new RedisArray($servers);

$iterator = null;
while ($members = $client->sScan('set', $iterator)) {
    foreach ($members as $member) {
        echo $member . PHP_EOL;
    }
}
```

Throws error 
```
Warning: Redis::sscan(): Argument #2 ($i_iterator) must be passed by reference, value given in /home/longman/projects/web/redis-test/public/redis-array.php on line 14
```

This works for `Redis` class.

PHP: 8.0.14
php-redis: 5.3.5

**Comment by akalongman** [comment]
It's a major blocker bug for us, please fix it ASAP :heart_eyes:

**Comment by akalongman** [comment]
Related issue https://github.com/phpredis/phpredis/issues/1393

**Comment by yatsukhnenko** [comment]
@akalongman but why do you pass `$iterator` by value and not by reference?

**Comment by akalongman** [comment]
@yatsukhnenko show me the example what do you mean

**Comment by yatsukhnenko** [comment]
```php
$client->sScan('set', &$iterator)
```

**Comment by akalongman** [comment]
@yatsukhnenko https://www.php.net/manual/en/language.references.pass.php

> Note: There is no reference sign on a function call - only on function definitions. Function definitions alone are enough to correctly pass the argument by reference.

Also, my example works when I use `Redis` class, instead of `RedisArray`

**Comment by yatsukhnenko** [comment]
The problem is that RedisArray doesn't have method sscan, it just forward call to Redis class with arguments passed as is. Try to make the reference first and then pass it to RedisArray::sscan and it should works
```php
$tmp = null;
$iterator = &$tmp;
$client->sScan('set', $iterator);
```

**Comment by akalongman** [comment]
@yatsukhnenko result is the same.

```php
<?php

ini_set('error_reporting', E_ALL);
ini_set('display_errors', 1);

$servers = [
    '127.0.0.1:6379',
    '127.0.0.1:6380',
];

$client = new RedisArray($servers);

$iterator2 = null;
$iterator = &$iterator2;
$client->sScan('set', $iterator);
```

> Warning: Redis::sscan(): Argument # 2 ($i_iterator) must be passed by reference, value given in /home/longman/projects/web/testing/redis-array.php on line 15

If RedisArray forwards calls to Redis class, why then using Redis class directly does not raise the same error?

**Comment by yatsukhnenko** [comment]
I don't know, I will try to figure it out this weekend

**Comment by yatsukhnenko** [comment]
If it works fine for Redis class as workaround you may use `RedisArray::_instance` to get the object and call sscan

**Comment by akalongman** [comment]
This call is used in the Laravel framework, so I am not able to change anything :(

If I can somehow help in fixing this, let me know

**Comment by yatsukhnenko** [comment]
It looks like PHP `__call` or `call_user_func_array` behavior

```php
class A {
    public function foo(&$ref) {
    }
}

class B {
    public function __call($name, $arguments) {
        call_user_func_array([new A(), $name], $arguments);
    }
}

$tmp = null;
$b = new B();
$b->foo($tmp);
```

**Comment by akalongman** [comment]
Maybe override these methods in PhpArray class and call something like a parent::method() ?

**Comment by yatsukhnenko** [comment]
> Maybe override these methods in PhpArray class and call something like a parent::method() ?

If laravel framework allows this you can try to use something like this
```php
class MyRedisArray extends RedisArray
{
    public function sScan($key, &$iterator, $pattern = '', $count = 0)
    {
        return $this->_instance($this->_host($key))->sScan($key, $iterator, $pattern, $count);
    }
}
```

**Comment by akalongman** [comment]
Can we do this in php-redis extension itself?

**Comment by yatsukhnenko** [comment]
Technically we can but I prefer to find what's wrong with `__call` and `call_user_function` and not to add wrappers that may be easily implemented in PHP-script

**Comment by akalongman** [comment]
There is no easy way to override code from the vendor folder and especially from the Laravel framework, so please assign this issue high priority.

**Comment by yatsukhnenko** [comment]
@akalongman could you test `issue-2055` branch?

**Comment by akalongman** [comment]
@yatsukhnenko fixed errors for `sScan`, `zScan`, `hScan`, but still exists for `scan`:

```php
<?php
ini_set('error_reporting', E_ALL);
ini_set('display_errors', 1);

$servers = [
    '127.0.0.1:6379',
    '127.0.0.1:6380',
];

$redis = new RedisArray($servers);

$iterator = null;
while (false !== ($keys = $redis->scan($iterator))) {
    foreach ($keys as $key) {
        echo $key . PHP_EOL;
    }
}
```

And error message is different:
```
Fatal error: RedisArray::__call(): Could not find key in /home/longman/projects/testing/redis-array.php on line 19
```

**Comment by yatsukhnenko** [comment]
@akalongman I've implemented `RedisArray::scan` in branch `issue-1393`

**Comment by akalongman** [comment]
@yatsukhnenko ok, I will test and give feedback in #1393 then.

I think this changes can be merged to the master

**Comment by akalongman** [comment]
When will this fix be released?

**Comment by michael-grunder** [comment]
I can prepare a release in the next day or two.  

@yatsukhnenko Any other commits we should include?

**Comment by akalongman** [comment]
Can you include fixes for this issue? https://github.com/phpredis/phpredis/issues/1393

**Comment by yatsukhnenko** [comment]
@michael-grunder @akalongman a fix for 1393 isn't ready yet. I think I can finish it this weekend and then it also may be included to the release

**Comment by akalongman** [comment]
I've tested develop for this issue and seems like works everything. Thank you :heart_eyes:

**Comment by yatsukhnenko** [comment]
Fixed in 5.3.7
